Subject: [PATCH] [BZ 161888] lsluns: document restriction to zfcp-only systems
From: Jens Remus <jremus@linux.vnet.ibm.com>

Description:  lsluns: Fix filter handling and documentation enhancements.
Symptom:      lsluns lists all LUNs discovered in the FC SAN despite user
              given filter(s) that do not match anything:
                # lsluns -c 0.0.5080
                    No valid combination found for adapter '0.0.5080'. Removing
                    from resource list.
                No valid parameters left, using all available resources in
                system.
                Scanning for LUNs on adapter 0.0.5090
                ...

              lsluns prints the message "No valid combination found for
              {adapter|port} '...'. Removing from resource list." for every
              adapter and port filter that does not contribute to the final
              filtered results.

              The formatting of the lsluns (8) man page is flawed.

              lsluns is used in unexpected or even unsupported ways.
Problem:      Scanning can be resource consumptive. So if a user already wants
              to filter, possibly to reduce resource consumption, he does not
              want to happen to scan everything and thus consume the worst case
              of resources.

              The message "No valid combination found for {adapter|port} '...'.
              Removing from resource list." is potentially confusing. It is
              unclear which combination is being referred to, especially if
              only one single adapter or port filter was specified. There is
              also no differentiation whether the denoted adapter or port
              exists for its own or not. It just does not exist in the final
              filtered results.

              The formatting of the lsluns (8) man page is flawed.

              The lsluns usage text and lsluns (8) man page lack the
              information on the intended use cases, requirements, and
              restrictions.
Solution:     Print a message and exit when all filters match nothing.

              Do not print confusing messages when a filter matches nothing.

              Fix man page formatting.

              Enhance usage text and man page. Clarify discovery use case,
              relation to NPIV and to zfcp auto LUN scan. Point out
              IBM Storwize configuration requirements. Document restriction to
              zfcp-only systems.
Reproduction: Use lsluns and specify adapter bus-ID and/or target port WWPN
              filter(s) that do not match anything. Either specify inexistent
              bus-IDs and/or WWPNs or invalid filter arguments.
Upstream-ID:  55f0b001d15b3984967e7f87d6118232078e3fe5
Problem-ID:   161888

Upstream-Description:

              lsluns: document restriction to zfcp-only systems

              Signed-off-by: Steffen Maier <maier@linux.vnet.ibm.com>
              Reviewed-by: Jens Remus <jremus@linux.vnet.ibm.com>
              Signed-off-by: Michael Holzheu <holzheu@linux.vnet.ibm.com>


Signed-off-by: Jens Remus <jremus@linux.vnet.ibm.com>
---
 zconf/lsluns   |    4 ++++
 zconf/lsluns.8 |    8 ++++++--
 2 files changed, 10 insertions(+), 2 deletions(-)

--- a/zconf/lsluns
+++ b/zconf/lsluns
@@ -170,6 +170,10 @@ sub get_lun_hash
 sub lsluns_usage {
     print <<EOD;
 Usage:
+This tool is designed for environments where all SCSI devices are attached
+through the zfcp device driver. Expect error messages in mixed environments
+such as with iSCSI.
+
 $PROGRAM_NAME [-c <busid>] ... [-p <wwpn>] ... [-h] [-v]
 
     List LUNs discovered in the Fibre Channel (FC) Storage Area Network (SAN).
--- a/zconf/lsluns.8
+++ b/zconf/lsluns.8
@@ -4,8 +4,8 @@
 .\"
 .TH LSLUNS 8 "2017-02-17" "s390-tools"
 .SH NAME
-lsluns \- list LUNs discovered in the FC SAN, or show encryption state of
-attached LUNs
+lsluns \- list LUNs discovered in the FC SAN through zfcp, or show encryption state of
+zfcp-attached LUNs
 
 .SH SYNOPSIS
 .B lsluns
@@ -28,6 +28,10 @@ attached LUNs
 
 .SH DESCRIPTION
 .PP
+This tool is designed for environments where all SCSI devices are attached
+through the zfcp device driver. Expect error messages in mixed environments
+such as with iSCSI.
+
 .B lsluns
 lists all logical unit numbers (LUNs) discovered in the
 Fibre Channel (FC) Storage Area Network (SAN).
